### DISCUSSION TITLE
Polyglot for macro generated messages

### DISCUSSION THREAD

**Discussion by napolitanod** [mclemente/fvtt-module-polyglot#170] [Ideas]
Hello!

Thought I'd reach out to see if you would be inclined to consider, if it is even possible, extending polyglot API or functionality so it can be tapped for chat messages and chat bubbles created within a macro or script (e.g. module). Below are two different calls to generate either a chat message or chat bubble and, based on my testing, these do not render in polyglot (and the chatbubble I get a console log error message for when sent over socket). 

My interest is from a module development standpoint, because it would be cool if a module that generated text could tap into polyglot, providing whatever flags or data may be needed to facilitate the call.

ChatMessage.create({
   speaker:game.user,
   content : "Hello Hello",
  ...whatever else module needs
},{chatBubble : false})

canvas.hud.bubbles.say()  

Thanks for your time!!!!

**Comment by mclemente** [comment]
I'm not sure how to handle bubbles, but I've added support to ChatMessage.create() with the new update you just gotta add the following now:
```js
ChatMessage.create({
speaker:game.user,
content : "Hello Hello",
type: CONST.CHAT_MESSAGE_TYPES.IC, //or 2. Can also be OOC (1) or WHISPER (4) based on the Scramble OOC Chat Messages setting
lang: "dwarvish" //or "Dwarvish". Optional, if not specified, it'll get the currently selected language on the selector.
},{chatBubble : false})
```
This should not affect most current functions/macros since they generally aren't written with a type.

**Reply by napolitanod** [reply]
Sorry for the delay. Thank you for responding! I'm going to given this a shot in my module and see if I can get the messaging to work!

Thanks!

**Reply by mclemente** [reply]
For `canvas.hud.bubbles.say()`, I'm looking into using the `emote` parameter as a way to pass options.
For example:

```js
async foo(options={}) {
  const emote = Object.keys(options).length ? {emote: options} : false;
  await canvas.hud.bubbles.say(token, result, emote);
}
foo({language: "dwarvish"});
```

**Reply by napolitanod** [reply]
Thank you for your response and consideration!

I had a chance to test out the ChatMessage and the in character translation generates great for text. I do see that when additional html is provided for chat message formatting it appears that this will be included in the translation. Please let me know if there is a setting I am missing that bypasses this. See image I included below of the first run through a message with html and the second the same message without the html. I'm guessing this happens because polyglot captures what would be the existing chat message and builds a new one around this, so what is intended to be html in the original becomes what is spoken in the new.  

I checked the polyglot API and didn't see something that can be used to leverage polyglot translation, but this might be the most flexible way (I say not knowing the polyglot code, so sorry if my ignorance is overlooking an impossibility). But thinking through, if a module could call polyglot, pass in the text for translation, and get back and object containing the language used and translation(if any), then that module could use the translation both in the chat bubble and to build a custom chat message around it. Does that at least sound like a feasible option?

For now I'm going to see if there's something I can do to capture the chat message created by polyglot due to the original message just as text via hook and add the additional html to that. I think that may work.

Thank you again,
- Dan

![image](https://user-images.githubusercontent.com/22696153/127165940-41c026e0-ccde-4232-a07a-29033fdaed30.png)

**Reply by mclemente** [reply]
Hm, it would be really "hacky". Maybe moving the link to a flag during preCreateChatMessage and then working on it later?
The best idea so far is using Regex to remove the link during renderChatMessage and then adding it back to the message later, but it isn't working exactly right now. I'll take a better look at it,

**Reply by napolitanod** [reply]
I got the chat bubble integrating with Token Says module now and think I have a workable solution for chat message by hooking on renderChatMessage and doing jquery there to insert the Token Says html into the polyglot (if it is there).

Thanks for your help!

- Dan

**Reply by mclemente** [reply]
I've finally found out there was a simple way to make Polyglot to ignore messages with links. All I had to do was a set a Hook to createChatMessage and make it return earlier if it started with a "<".
